File: src/pricing/claude.ts
// Per-model Claude pricing from https://platform.claude.com/docs/en/about-claude/pricing
// Cache read = 0.10x input; cache create (5min) = 1.25x input.
// Keep entries keyed by normalized family-first id (with date/latest suffixes stripped).
// Last updated: 2026-08-03. Run `pnpm tsx scripts/update-pricing.ts` to check for drift.

import { CACHE_READ_RATE_MULTIPLIER } from '../constants.js';
import { stripModelVersionSuffixes } from '../data/modelId.js';
import type { FallbackCollector } from './fallback.js';

export interface ClaudePricing {
  inputPerMillion: number;
  outputPerMillion: number;
  cacheReadPerMillion: number;
  cacheCreatePerMillion: number;
}

function priceFromBase(input: number, output: number): ClaudePricing {
  return {
    inputPerMillion: input,
    outputPerMillion: output,
    cacheReadPerMillion: input * CACHE_READ_RATE_MULTIPLIER,
    cacheCreatePerMillion: input * 1.25,
  };
}

export const CLAUDE_PRICING_BY_ID: Record<string, ClaudePricing> = {
  // Latest generation (Fable/Mythos tier: $10/$50)
  'claude-fable-5': priceFromBase(10, 50),
  'claude-mythos-5': priceFromBase(10, 50),
  // Current generation (Opus 4.5+ and Sonnet/Haiku 4.5+ tier)
  'claude-opus-5': priceFromBase(5, 25),
  'claude-opus-4-8': priceFromBase(5, 25),
  'claude-opus-4-7': priceFromBase(5, 25),
  'claude-opus-4-6': priceFromBase(5, 25),
  'claude-opus-4-5': priceFromBase(5, 25),
  'claude-sonnet-4-6': priceFromBase(3, 15),
  'claude-sonnet-4-5': priceFromBase(3, 15),
  'claude-sonnet-5': priceFromBase(3, 15),
  'claude-haiku-4-5': priceFromBase(1, 5),
  // Older / deprecated
  'claude-opus-4-1': priceFromBase(15, 75),
  'claude-opus-4-0': priceFromBase(15, 75),
  'claude-opus-4': priceFromBase(15, 75),
  'claude-sonnet-4-0': priceFromBase(3, 15),
  'claude-sonnet-4': priceFromBase(3, 15),
  'claude-haiku-3-5': priceFromBase(0.8, 4),
  // Claude 3 generation. Listed explicitly because the family fallback prices
  // them at the modern tier, which is 3x too low for opus and 4x too high for
  // haiku, and the wrong figure gets persisted as costUSD.
  'claude-opus-3': priceFromBase(15, 75),
  'claude-sonnet-3-7': priceFromBase(3, 15),
  'claude-sonnet-3-5': priceFromBase(3, 15),
  'claude-sonnet-3': priceFromBase(3, 15),
  'claude-haiku-3': priceFromBase(0.25, 1.25),
};

// Historical pricing overrides. Each entry means "for usage dates strictly
// BEFORE `before`, this model was priced as `pricing` instead of the entry
// in CLAUDE_PRICING_BY_ID". Sort each list ascending by `before`.
//
// Anthropic normally ships a new model id at a new tier rather than re-pricing
// an existing one, so this table stays small. It exists for the exception:
// claude-sonnet-5 launched at introductory rates and moved to the standard
// tier later, and without the entry every day of sonnet-5 usage recorded
// before that date would be silently re-costed at the higher rate.
export const CLAUDE_PRICING_OVERRIDES: Record<
  string,
  Array<{ before: string; pricing: ClaudePricing }>
> = {
  // Introductory pricing through 2026-08-31; standard tier from 2026-09-01.
  'claude-sonnet-5': [{ before: '2026-09-01', pricing: priceFromBase(2, 10) }],
};

// Family fallback for unknown future models.
const FAMILY_FALLBACK: Record<'fable' | 'mythos' | 'opus' | 'sonnet' | 'haiku', ClaudePricing> = {
  fable: priceFromBase(10, 50),
  mythos: priceFromBase(10, 50),
  opus: priceFromBase(5, 25),
  sonnet: priceFromBase(3, 15),
  haiku: priceFromBase(1, 5),
};

// Called once per assistant entry while reading the JSONL corpus, which holds
// only a handful of distinct model strings, so the lowercase plus two regexes
// are worth caching.
const canonicalIdCache = new Map<string, string>();

function canonicalClaudeModelId(model: string): string {
  const cached = canonicalIdCache.get(model);
  if (cached !== undefined) return cached;

  const id = stripModelVersionSuffixes(model.toLowerCase());
  const legacy = /^claude-(\d+)(?:-(\d+))?-(opus|sonnet|haiku)$/.exec(id);
  const [, major, minor, family] = legacy ?? [];
  const canonical =
    legacy && major && family ? `claude-${family}-${major}${minor ? `-${minor}` : ''}` : id;

  canonicalIdCache.set(model, canonical);
  return canonical;
}

export function findClaudePricing(
  model: string,
  usageDate?: string,
  fallbacks?: FallbackCollector,
): ClaudePricing {
  const id = canonicalClaudeModelId(model);
  if (usageDate) {
    const overrides = CLAUDE_PRICING_OVERRIDES[id];
    if (overrides) {
      for (const entry of overrides) {
        if (usageDate < entry.before) return entry.pricing;
      }
    }
  }
  const exact = CLAUDE_PRICING_BY_ID[id];
  if (exact) return exact;
  for (const family of ['fable', 'mythos', 'opus', 'haiku', 'sonnet'] as const) {
    if (!id.includes(family)) {
      continue;
    }

    fallbacks?.record(id);
    return FAMILY_FALLBACK[family];
  }
  // Unrecognized entirely — still a guess, so it counts as a fallback hit.
  fallbacks?.record(id);
  return FAMILY_FALLBACK.sonnet;
}

/**
 * The one weighted-sum formula behind all three estimators below, which
 * previously each spelled out the same `(a*in + b*out + …)/1_000_000`.
 */
function claudeCost(
  pricing: ClaudePricing,
  tokens: { raw: number; output: number; cacheRead: number; cacheCreate: number },
): number {
  return (
    (tokens.raw * pricing.inputPerMillion +
      tokens.output * pricing.outputPerMillion +
      tokens.cacheRead * pricing.cacheReadPerMillion +
      tokens.cacheCreate * pricing.cacheCreatePerMillion) /
    1_000_000
  );
}

export interface ClaudeMessageUsage {
  input_tokens?: number;
  cache_read_input_tokens?: number;
  output_tokens?: number;
  cache_creation_input_tokens?: number;
}

export function estimateClaudeCostUSD(
  model: string,
  usage: ClaudeMessageUsage,
  usageDate?: string,
  fallbacks?: FallbackCollector,
): number {
  return claudeCost(findClaudePricing(model, usageDate, fallbacks), {
    raw: usage.input_tokens ?? 0,
    output: usage.output_tokens ?? 0,
    cacheRead: usage.cache_read_input_tokens ?? 0,
    cacheCreate: usage.cache_creation_input_tokens ?? 0,
  });
}

// Backfill estimator for synced rows that lack a costUSD value (older data).
// The cache vs raw-input split has already been collapsed into a single
// inputTokens number, so we apply full input pricing — an upper bound.
export function estimateClaudeCostFromAggregateTokens(
  model: string,
  inputTokens: number,
  outputTokens: number,
  usageDate?: string,
  fallbacks?: FallbackCollector,
): number {
  return claudeCost(findClaudePricing(model, usageDate, fallbacks), {
    raw: inputTokens,
    output: outputTokens,
    cacheRead: 0,
    cacheCreate: 0,
  });
}

export function claudeCountsHaveCostBreakdown(counts: {
  rawInputTokens?: number;
  cachedInputTokens?: number;
  cacheCreationInputTokens?: number;
}): boolean {
  return (
    counts.rawInputTokens !== undefined ||
    counts.cachedInputTokens !== undefined ||
    counts.cacheCreationInputTokens !== undefined
  );
}

export function estimateClaudeCostFromStoredCounts(
  model: string,
  counts: {
    inputTokens: number;
    outputTokens: number;
    rawInputTokens?: number;
    cachedInputTokens?: number;
    cacheCreationInputTokens?: number;
  },
  usageDate?: string,
  fallbacks?: FallbackCollector,
): number | undefined {
  if (!claudeCountsHaveCostBreakdown(counts)) return undefined;
  const cacheRead = counts.cachedInputTokens ?? 0;
  const cacheCreate = counts.cacheCreationInputTokens ?? 0;
  return claudeCost(findClaudePricing(model, usageDate, fallbacks), {
    raw: counts.rawInputTokens ?? Math.max(0, counts.inputTokens - cacheRead - cacheCreate),
    output: counts.outputTokens,
    cacheRead,
    cacheCreate,
  });
}

File: crates/aitrack-core/src/pricing/claude.rs
//! Claude pricing tables and cost estimation.

use super::fallback::FallbackCollector;
use crate::constants::CACHE_READ_RATE_MULTIPLIER;
use crate::data::model_id::{CLAUDE_FAMILIES, canonicalize_claude_model_id};
use std::collections::HashMap;
use std::sync::LazyLock;

#[derive(Debug, Clone, Copy)]
pub struct ClaudePricing {
    pub input_per_million: f64,
    pub output_per_million: f64,
    pub cache_read_per_million: f64,
    pub cache_create_per_million: f64,
}

const REDUCED_CACHE_READ_RATE_MULTIPLIER: f64 = 0.025;

fn price_from_base(input: f64, output: f64, cache_read_rate_multiplier: f64) -> ClaudePricing {
    ClaudePricing {
        input_per_million: input,
        output_per_million: output,
        cache_read_per_million: input * cache_read_rate_multiplier,
        cache_create_per_million: input * 1.25,
    }
}

static CLAUDE_PRICING_BY_ID: LazyLock<HashMap<&'static str, ClaudePricing>> = LazyLock::new(|| {
    let mut map = HashMap::new();

    // Latest generation (Fable/Mythos tier)
    map.insert(
        "claude-fable-5-1",
        price_from_base(10.0, 50.0, REDUCED_CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "claude-mythos-5-1",
        price_from_base(10.0, 50.0, REDUCED_CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "claude-fable-5",
        price_from_base(10.0, 50.0, CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "claude-mythos-5",
        price_from_base(10.0, 50.0, CACHE_READ_RATE_MULTIPLIER),
    );

    // Current generation
    map.insert(
        "claude-opus-5",
        price_from_base(5.0, 25.0, CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "claude-opus-4-8",
        price_from_base(5.0, 25.0, CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "claude-opus-4-7",
        price_from_base(5.0, 25.0, CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "claude-opus-4-6",
        price_from_base(5.0, 25.0, CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "claude-opus-4-5",
        price_from_base(5.0, 25.0, CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "claude-sonnet-4-6",
        price_from_base(3.0, 15.0, CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "claude-sonnet-4-5",
        price_from_base(3.0, 15.0, CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "claude-sonnet-5",
        price_from_base(2.0, 10.0, CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "claude-haiku-4-5",
        price_from_base(1.0, 5.0, CACHE_READ_RATE_MULTIPLIER),
    );

    // Older / deprecated
    map.insert(
        "claude-opus-4-1",
        price_from_base(15.0, 75.0, CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "claude-opus-4-0",
        price_from_base(15.0, 75.0, CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "claude-opus-4",
        price_from_base(15.0, 75.0, CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "claude-sonnet-4-0",
        price_from_base(3.0, 15.0, CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "claude-sonnet-4",
        price_from_base(3.0, 15.0, CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "claude-haiku-3-5",
        price_from_base(0.8, 4.0, CACHE_READ_RATE_MULTIPLIER),
    );

    // Claude 3 generation
    map.insert(
        "claude-opus-3",
        price_from_base(15.0, 75.0, CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "claude-sonnet-3-7",
        price_from_base(3.0, 15.0, CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "claude-sonnet-3-5",
        price_from_base(3.0, 15.0, CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "claude-sonnet-3",
        price_from_base(3.0, 15.0, CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "claude-haiku-3",
        price_from_base(0.25, 1.25, CACHE_READ_RATE_MULTIPLIER),
    );

    map
});

static FAMILY_FALLBACK: LazyLock<HashMap<&'static str, ClaudePricing>> = LazyLock::new(|| {
    let mut map = HashMap::new();
    map.insert(
        "fable",
        price_from_base(10.0, 50.0, CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "mythos",
        price_from_base(10.0, 50.0, CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "opus",
        price_from_base(5.0, 25.0, CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "sonnet",
        price_from_base(3.0, 15.0, CACHE_READ_RATE_MULTIPLIER),
    );
    map.insert(
        "haiku",
        price_from_base(1.0, 5.0, CACHE_READ_RATE_MULTIPLIER),
    );
    map
});

pub fn find_claude_pricing(
    model: &str,
    _usage_date: Option<&str>,
    fallbacks: Option<&FallbackCollector>,
) -> ClaudePricing {
    let id = canonicalize_claude_model_id(model);

    if let Some(pricing) = CLAUDE_PRICING_BY_ID.get(id.as_str()) {
        return *pricing;
    }

    // Try family fallback
    for family in CLAUDE_FAMILIES {
        if id.contains(family) {
            if let Some(fb) = fallbacks {
                fb.record(&id);
            }
            return *FAMILY_FALLBACK.get(family).unwrap();
        }
    }

    // Unrecognized - use sonnet fallback
    if let Some(fb) = fallbacks {
        fb.record(&id);
    }
    *FAMILY_FALLBACK.get("sonnet").unwrap()
}

fn claude_cost(pricing: ClaudePricing, tokens: &ClaudeTokens) -> f64 {
    (tokens.raw * pricing.input_per_million
        + tokens.output * pricing.output_per_million
        + tokens.cache_read * pricing.cache_read_per_million
        + tokens.cache_create * pricing.cache_create_per_million)
        / 1_000_000.0
}

#[derive(Debug, Clone, Copy)]
struct ClaudeTokens {
    raw: f64,
    output: f64,
    cache_read: f64,
    cache_create: f64,
}

#[derive(Debug, Clone, Copy, Default)]
pub struct ClaudeMessageUsage {
    pub input_tokens: Option<u64>,
    pub cache_read_input_tokens: Option<u64>,
    pub output_tokens: Option<u64>,
    pub cache_creation_input_tokens: Option<u64>,
}

pub fn estimate_claude_cost_usd(
    model: &str,
    usage: &ClaudeMessageUsage,
    usage_date: Option<&str>,
    fallbacks: Option<&FallbackCollector>,
) -> f64 {
    let pricing = find_claude_pricing(model, usage_date, fallbacks);
    claude_cost(
        pricing,
        &ClaudeTokens {
            raw: usage.input_tokens.unwrap_or(0) as f64,
            output: usage.output_tokens.unwrap_or(0) as f64,
            cache_read: usage.cache_read_input_tokens.unwrap_or(0) as f64,
            cache_create: usage.cache_creation_input_tokens.unwrap_or(0) as f64,
        },
    )
}

pub fn estimate_claude_cost_from_aggregate_tokens(
    model: &str,
    input_tokens: u64,
    output_tokens: u64,
    usage_date: Option<&str>,
    fallbacks: Option<&FallbackCollector>,
) -> f64 {
    let pricing = find_claude_pricing(model, usage_date, fallbacks);
    claude_cost(
        pricing,
        &ClaudeTokens {
            raw: input_tokens as f64,
            output: output_tokens as f64,
            cache_read: 0.0,
            cache_create: 0.0,
        },
    )
}

pub fn claude_counts_have_cost_breakdown(counts: &ClaudeCountsBreakdown) -> bool {
    counts.raw_input_tokens.is_some()
        || counts.cached_input_tokens.is_some()
        || counts.cache_creation_input_tokens.is_some()
}

#[derive(Debug, Clone, Copy, Default)]
pub struct ClaudeCountsBreakdown {
    pub input_tokens: u64,
    pub output_tokens: u64,
    pub raw_input_tokens: Option<u64>,
    pub cached_input_tokens: Option<u64>,
    pub cache_creation_input_tokens: Option<u64>,
}

pub fn estimate_claude_cost_from_stored_counts(
    model: &str,
    counts: &ClaudeCountsBreakdown,
    usage_date: Option<&str>,
    fallbacks: Option<&FallbackCollector>,
) -> Option<f64> {
    if !claude_counts_have_cost_breakdown(counts) {
        return None;
    }

    let cache_read = counts.cached_input_tokens.unwrap_or(0);
    let cache_create = counts.cache_creation_input_tokens.unwrap_or(0);
    let raw = counts.raw_input_tokens.unwrap_or_else(|| {
        counts
            .input_tokens
            .saturating_sub(cache_read + cache_create)
    });

    let pricing = find_claude_pricing(model, usage_date, fallbacks);
    Some(claude_cost(
        pricing,
        &ClaudeTokens {
            raw: raw as f64,
            output: counts.output_tokens as f64,
            cache_read: cache_read as f64,
            cache_create: cache_create as f64,
        },
    ))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_find_claude_pricing() {
        let pricing = find_claude_pricing("claude-opus-4-6", None, None);
        assert_eq!(pricing.input_per_million, 5.0);
        assert_eq!(pricing.output_per_million, 25.0);
    }

    #[test]
    fn test_estimate_claude_cost() {
        let usage = ClaudeMessageUsage {
            input_tokens: Some(1_000_000),
            output_tokens: Some(1_000_000),
            ..Default::default()
        };
        let cost = estimate_claude_cost_usd("claude-opus-4-6", &usage, None, None);
        assert_eq!(cost, 30.0); // 5 + 25
    }
}
